nomad/stream/event_publisher.go

```diff
@@ -0,0 +1,209 @@
+package stream
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/nomad/nomad/structs"
+
+	"github.com/hashicorp/go-hclog"
+)
+
+const (
+	DefaultTTL = 1 * time.Hour
+)
+
+type EventPublisherCfg struct {
+	EventBufferSize int64
+	EventBufferTTL  time.Duration
+	Logger          hclog.Logger
+}
+
+type EventPublisher struct {
+	// lock protects the eventbuffer
+	lock sync.Mutex
+
+	// eventBuf stores a configurable amount of events in memory
+	eventBuf *eventBuffer
+
+	// pruneTick is the duration to periodically prune events from the event
+	// buffer. Defaults to 5s
+	pruneTick time.Duration
+
+	logger hclog.Logger
+
+	subscriptions *subscriptions
+
+	// publishCh is used to send messages from an active txn to a goroutine which
+	// publishes events, so that publishing can happen asynchronously from
+	// the Commit call in the FSM hot path.
+	publishCh chan structs.Events
+}
+
+type subscriptions struct {
+	// lock for byToken. If both subscription.lock and EventPublisher.lock need
+	// to be held, EventPublisher.lock MUST always be acquired first.
+	lock sync.RWMutex
+
+	// byToken is an mapping of active Subscriptions indexed by a token and
+	// a pointer to the request.
+	// When the token is modified all subscriptions under that token will be
+	// reloaded.
+	// A subscription may be unsubscribed by using the pointer to the request.
+	byToken map[string]map[*SubscribeRequest]*Subscription
+}
+
+func NewEventPublisher(ctx context.Context, cfg EventPublisherCfg) *EventPublisher {
+	if cfg.EventBufferTTL == 0 {
+		cfg.EventBufferTTL = 1 * time.Hour
+	}
+
+	if cfg.Logger == nil {
+		cfg.Logger = hclog.NewNullLogger()
+	}
+
+	buffer := newEventBuffer(cfg.EventBufferSize, cfg.EventBufferTTL)
+	e := &EventPublisher{
+		logger:    cfg.Logger.Named("event_publisher"),
+		eventBuf:  buffer,
+		publishCh: make(chan structs.Events, 64),
+		subscriptions: &subscriptions{
+			byToken: make(map[string]map[*SubscribeRequest]*Subscription),
+		},
+		pruneTick: 5 * time.Second,
+	}
+
+	go e.handleUpdates(ctx)
+	go e.periodicPrune(ctx)
+
+	return e
+}
+
+func (e *EventPublisher) Len() int {
+	return e.eventBuf.Len()
+}
+
+// Publish events to all subscribers of the event Topic.
+func (e *EventPublisher) Publish(events structs.Events) {
+	if len(events.Events) > 0 {
+		e.publishCh <- events
+	}
+}
```

Comment:
Since events are readonly why not use a pointer to save copying the struct repeatedly?
```suggestion
// Publish events to all subscribers of the event Topic.
func (e *EventPublisher) Publish(events *structs.Events) {
	if len(events.Events) > 0 {
		e.publishCh <- events
	}
}
```